internal/security/proxy/service.go

```diff
@@ -0,0 +1,332 @@
+ * the License.
+ *
+ * @author: Tingyu Zeng, Dell
+ * @version: 1.1.0
+ *******************************************************************************/
+package proxy
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/dghubble/sling"
+	"net/http"
+)
+
+type Service struct {
+	Connect    Requestor
+	CertCfg    CertConfig
+	ServiceCfg ServiceConfig
+}
+
+type ServiceConfig interface {
+	GetProxyAuthMethod() string
+	GetProxyAuthTTL() int
+	GetProxyAuthResource() string
+	GetProxyACLName() string
+	GetProxyACLWhiteList() string
+	GetSecretSvcSNIS() string
+	GetEdgeXSvcs() map[string]service
+}
+
+func (s *Service) CheckProxyServiceStatus() error {
+	return s.checkServiceStatus(s.Connect.GetProxyBaseURL())
+}
+
+func (s *Service) CheckSecretServiceStatus() error {
+	return s.checkServiceStatus(s.Connect.GetSecretSvcBaseURL())
+}
+
+func (s *Service) checkServiceStatus(path string) error {
+	req, err := sling.New().Get(path).Request()
+	resp, err := s.Connect.GetHTTPClient().Do(req)
+	if err != nil {
+		e := fmt.Sprintf("the status of service on %s is unknown, the initialization is terminated", path)
+		return errors.New(e)
+	}
+	defer resp.Body.Close()
+
+	switch resp.StatusCode {
+	case http.StatusOK:
+		lc.Info(fmt.Sprintf("the service on %s is up successfully", path))
+		break
+	default:
+		e := fmt.Sprintf("the service on %s is down", path)
+		return errors.New(e)
+	}
+	return nil
+}
+
+func (s *Service) ResetProxy() error {
+	paths := []string{RoutesPath, ServicesPath, ConsumersPath, PluginsPath, CertificatesPath}
+	for _, path := range paths {
+		d, err := s.getSvcIDs(path)
+		if err != nil {
+			return err
+		}
+		for _, c := range d.Section {
+			r := &Resource{c.ID, s.Connect}
+			err = r.Remove(path)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (s *Service) Init() error {
+	err := s.loadCert()
+	if err != nil {
+		return err
+	}
+
+	for _, service := range s.ServiceCfg.GetEdgeXSvcs() {
+		serviceParams := &KongService{
+			Name:     service.Name,
+			Host:     service.Host,
+			Port:     service.Port,
+			Protocol: service.Protocol,
+		}
+
+		err := s.initKongService(serviceParams)
+		if err != nil {
+			return err
+		}
+
+		routeParams := &KongRoute{
+			Paths: []string{"/" + service.Name},
+			Name:  service.Name,
+		}
+		err = s.initKongRoutes(routeParams, service.Name)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = s.initAuthmethod(s.ServiceCfg.GetProxyAuthMethod(), s.ServiceCfg.GetProxyAuthTTL())
+	if err != nil {
+		return err
+	}
+
+	err = s.initACL(s.ServiceCfg.GetProxyACLName(), s.ServiceCfg.GetProxyACLWhiteList())
+	if err != nil {
+		return err
+	}
+
+	lc.Info("finishing initialization for reverse proxy")
+	return nil
+}
+
+func (s *Service) loadCert() error {
+	cp, err := s.getCertPair()
+	if err != nil {
+		return err
+	}
+	body := &CertInfo{
+		Cert: cp.Cert,
+		Key:  cp.Key,
+		Snis: []string{s.ServiceCfg.GetSecretSvcSNIS()},
```

Comment:
the requirement from KONG is to accept a list so a single string will get error code 400.

Reply:
No, I'm saying that the configuration.toml only allows a single string for the SNI setting, but in the code we transform the single SNI into a list because Kong requires a list, but we could have just as easily required that there be a list of SNIs in the configuration.toml and then we wouldn't have needed to modify the single string into a list of 1 string.

This would allow the user to specify multiple SNI's for Kong proxies. 

Please file an issue for this

Reply:
tracked under #1546